lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -324,6 +354,604 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      if (WinFSO.has_value() && (isa<ConstantInt>(CI->getArgOperand(1)))
+          && (cast<ConstantInt>(CI->getArgOperand(1))->getZExtValue()
+              == *WinFSO))
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+    Type = FunctionTypeValue::NoReturn;
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Union between callee-saved registers and SP.
+  std::set<GlobalVariable *> CSAndSPRegs = { CalleeSavedRegs };
+  CSAndSPRegs.insert(GCBI->spReg());
+
+  // Refine ABI analyses results.
+  for (const auto &Reg : CSAndSPRegs) {
+    if (ABIResults.Arguments.count(Reg) != 0)
+      ABIResults.Arguments[Reg] = model::RegisterState::Values::No;
+  }
+
+  for (const auto &[K, _] : ABIResults.CallSites) {
+    for (const auto &Reg : CSAndSPRegs) {
+      if (ABIResults.CallSites[K].Arguments.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .Arguments[Reg] = model::RegisterState::Values::No;
+
+      if (ABIResults.CallSites[K].ReturnValues.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .ReturnValues[Reg] = model::RegisterState::Values::No;
+    }
+  }
+
+  for (const auto &E : Result) {
+    if (E.second != FunctionEdgeTypeValue::Return) {
```

Comment:
`IndirectTailCall` is a valid return too I'd say.

Reply:
Right!